### PR TITLE
Products

### PR DESCRIPTION
**Pull request by vinc3nati**
Completed Product filters
- sorting
- price slider
- category filter
- rating filter
image

### CODE REVIEW

**Review by shrey27 — APPROVED** [review]
- overall code structure looks good
- file structure is also easily understandable
- PR looks good to merge

File: src/components/loader/loader.css [hunk 1]
```diff
@@ -0,0 +1,112 @@
+.container {
+  width: 100vw;
+  height: 100vh;
+  top: 0;
+  left: 0;
+  position: fixed;
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  background: rgba(0, 0, 0, 0.7);
```
Comment (PeeyushPrashant) [inline src/components/loader/loader.css:10]: try using color variables, don't hardcode it

File: src/contexts/product-context.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          dispatch({
+            type: ACTIONS.GetProducts,
+            payload: { products: prodResponse.data.products },
+          });
+        }
+
+        const categoriesResponse = await getCategories();
+        if (categoriesResponse.data.categories) {
+          dispatch({
+            type: ACTIONS.GetCategories,
```
Comment (PeeyushPrashant) [inline src/contexts/product-context.js:33]: liked this way of keeping the type attribute in capital

**Review by PeeyushPrashant — APPROVED** [review]
liked your folder structure.
code looks very clean

File: src/components/loader/loader.css (hunk 1 above)
Comment (vinc3nati) [reply src/components/loader/loader.css:10]: sure

File: src/App.js [hunk 3]
```diff
@@ -6,16 +6,23 @@ import { Home } from "./pages/Home/Home";
...
 function App() {
+  const { loader } = useData();
   return (
     <>
+      {loader && <Loader />}
       <Navbar />
       <Routes>
         <Route exact path="/" element={<Home />} />
         <Route exact path="/products" element={<Products />} />
+        {/* <Route path="/product/:productId" element={<ProductDetails />} /> */}
```
Comment (Shah-Arsalan) [inline src/App.js:22]: remove comments

File: src/components/product/Filters.jsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  const maxValue = state.products.reduce(
+    (acc, curr) => (Number(curr.price) > acc ? Number(curr.price) : acc),
+    0
+  );
+  return (
+    <aside className="product-aside">
+      <div className="aside-heading">
+        <header className="aside-header">Filter</header>
+        <span
+          style={{ cursor: "pointer" }}
```
Comment (Shah-Arsalan) [inline src/components/product/Filters.jsx:18]: avoid inline styling

File: src/components/product/Filters.jsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        <input
+          type="range"
+          name="priceRange"
+          className="slider"
+          min="0"
+          max={maxValue}
+          value={state.filters.priceRange}
+          onChange={(e) =>
+            dispatch({
+              type: ACTIONS.ChangeFilters,
```
Comment (Shah-Arsalan) [inline src/components/product/Filters.jsx:62]: loved the use of variables!

File: src/hooks/FilteredData.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  searchProducts,
+  filterByCategory,
+  filterByPrice,
+  filterByRating,
+  sortProducts,
+} from "../utilities/productFilters";
+
+const union = (...arr) =>
+  arr.reduce((acc, curr) =>
+    //   acc.concat(curr.filter((el) => !acc.some((ele) => ele.id === el.id))),
```
Comment (Shah-Arsalan) [inline src/hooks/FilteredData.js:13]: this one too

**Review by Shah-Arsalan — APPROVED** [review]

**Review by mdarif6 — APPROVED** [review]
-it is looking good to kept ACTION in object.
-folder structure is looking clean.

File: src/App.js (hunk 3 above)
Comment (Rohanmond) [inline src/App.js:22]: remove unwanted comment

**Review by Rohanmond — APPROVED** [review]
looks good to merge

**Review by anuj-kumary — APPROVED** [review]

**Pull request merged**